woocommerce/woocommerce-android
# Refactor ProductUI to OrderCreationProduct

Part of: #9541

### Why
As part of the support for extensions milestone 2, we are working on support bundle products in the order creation flow.

### Description
This PR refactors and introduces some core changes needed to support bundles in the order creation flow. 
First, it introduces the  `ProductRules` and `ProductConfiguration` concepts.  Product rules will represent the settings a configurable product can have, like the min and max quantities or whether the product is optional. The configuration will represent a custom configuration these rules can have, like setting the current quantity of a product with min/max rules.

This code also introduces the OrderCreationProduct class. Our current representation of `Order.Items` it is plain, which makes it hard to work with parent/child relationships. This new OrderCreationProduct will be helpful when working with grouped products because, from the parent, we will have access to the children. It also adds new field rules so we can differentiate between products that need a configuration and products that don't

⚠️ This PR is not functional because it first needs to solve some conflicts with the `OrderCreationProduct` class. Because it was getting bigger, I decided to solve those conflicts in a new PR here https://github.com/woocommerce/woocommerce-android/pull/9850

Once the PR solving the conflicts is merged into this branch, you can test the basic navigation to the configuration screen when adding a configurable product to an order

### Testing instructions

#### Prerequisites 

1.  Install the product bundles extension
2. Create a bundle product requiring a configuration

#### Testing
TC1
1. Go to the order list screen
2. Tap on add new order (+)
3. Tap on Add products
4. Select a bundle product
5. Check that the app navigates to the configuration screen

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/18119390/505f8447-94f9-4112-a207-0c8fa8f3e43b
